File: gtci/00-warmup/cpp/4-is-anagram.cpp
#include <iostream>
#include <unordered_map>

class Solution {
   public:
    bool isAnagram(std::string s, std::string t) {
        if (s.size() != t.size()) return false;

        std::unordered_map<char, int> char_counts{};

        for (int i = 0; i < s.size(); i++) {
            char_counts[s[i]]++;
            char_counts[t[i]]--;
        }

        for (const std::pair<const char, int>& kv : char_counts) {
            if (kv.second != 0) {
                return false;
            }
        }

        return true;
    }
};

#include <cassert>

int main() {
    Solution sol;
    assert(sol.isAnagram("listen", "silent") == true);
    assert(sol.isAnagram("hello", "world") == false);
    assert(sol.isAnagram("anagram", "nagaram") == true);
    assert(sol.isAnagram("rat", "car") == false);
    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/00-warmup/python3/4-is-anagram.py
def isAnagram(s, t):
    if len(s) != len(t):
        return False

    char_counts = {}
    for ch in s:
        if ch not in char_counts:
            char_counts[ch] = 0
        char_counts[ch] += 1

    for ch in t:
        if ch not in char_counts:
            return False

        char_counts[ch] -= 1

    for _, count in char_counts.items():
        if count != 0:
            return False

    return True


if __name__ == "__main__":
    assert isAnagram("listen", "silent") == True
    assert isAnagram("hello", "world") == False
    assert isAnagram("anagram", "nagaram") == True
    assert isAnagram("rat", "car") == False
    print("All test cases passed.")
